### DISCUSSION TITLE
Can Flet run any Python code?

### DISCUSSION THREAD

**Discussion by ghost** [flet-dev/flet#220] [Q&A]
For example, is it possible to run numpy or a custom library with a compiled module?
Would it be possible to include PyQT widgets?

Also, will mobile apps be able to run Python code, just like Desktop (or web) apps?

Finally, how to include custom Flutter widgets to Flet?

**Comment by FeodorFitsner** [comment]
Thanks for your interest!

You can run any Python code as it's a regular Python code running on server-side (for now).

For mobile I think we will be supporting two modes: 1) server-driven UI when Python code is running on the server and updates UI on mobile via WebSocket and 2) when Python code runs on a mobile device - this case we should see if a certain Python library with native code could be run on a mobile. "Pure" Python libraries are going to be fine. Take a look at this doc for more details on Flet mobile strategy: https://flet.dev/blog/flet-mobile-strategy

Regarding custom widget - hop to a discussion forum here: https://discord.com/channels/981374556059086931/1011353284449157161

I'm going to create a guide on Flet controls development.

**Reply by ghost** [reply]
Could Flet be used to create a collaborative web site app that looks and behave like an app, with desktop and/or mobile clients that have complementary capabilities (specific to their platform), all developed and maintained with a common code base? Is the WebSocket protocol reliable enough to support such a distributed/collaborative application?

**Reply by FeodorFitsner** [reply]
Collaborative apps is where Flet really shines! Look at this (yet unfinished) tutorial for a chat application: https://github.com/flet-dev/website/blob/chat-tutorial/docs/tutorials/python-realtime-chat.md
Very soon, when `FletApp` widget for Flutter is published you'll be able to build a mobile white-labeled client for your app. Look, people are already tried using the widget on Android and it worked: https://github.com/flet-dev/flet/issues/213
WebSocket implementation in Flet is reliable - there are multiple safeguards and re-connecting logic on server and client sides.

**Reply by ghost** [reply]
Nice tutorial, but the demo on flet-chat.fly.dev does not seem to be up (or my DNS does not find it).
Edit: I see, it's in the making... I can't wait to try it! :+1:
